drivers/sensor/pcnt_esp32/pcnt_esp32.c

```diff
@@ -0,0 +1,415 @@
+
+	return 0;
+}
+#endif /* CONFIG_PCNT_ESP32_TRIGGER */
+
+static const struct sensor_driver_api pcnt_esp32_api = {
+	.sample_fetch = pcnt_esp32_sample_fetch,
+	.channel_get = pcnt_esp32_channel_get,
+	.attr_set = pcnt_esp32_attr_set,
+	.attr_get = pcnt_esp32_attr_get,
+#ifdef CONFIG_PCNT_ESP32_TRIGGER
+	.trigger_set = pcnt_esp32_trigger_set,
+#endif /* CONFIG_PCNT_ESP32_TRIGGER */
+};
+
+PINCTRL_DT_INST_DEFINE(0);
+
+#define UNIT_CONFIG(node_id)                                                                       \
+	{                                                                                          \
+		.idx = DT_REG_ADDR(node_id),                                                       \
+		.filter = DT_PROP_OR(node_id, filter, 0) > 1024 ? 1024                             \
+								: DT_PROP_OR(node_id, filter, 0),  \
+		.channel_config[0] =                                                               \
+			{                                                                          \
+				.sig_pos_mode = DT_PROP_OR(DT_CHILD(node_id, channela_0),          \
+							   sig_pos_mode, 0),                       \
+				.sig_neg_mode = DT_PROP_OR(DT_CHILD(node_id, channela_0),          \
+							   sig_neg_mode, 0),                       \
+				.ctrl_l_mode =                                                     \
+					DT_PROP_OR(DT_CHILD(node_id, channela_0), ctrl_l_mode, 0), \
+				.ctrl_h_mode =                                                     \
+					DT_PROP_OR(DT_CHILD(node_id, channela_0), ctrl_h_mode, 0), \
+			},                                                                         \
+		.channel_config[1] =                                                               \
+			{                                                                          \
+				.sig_pos_mode = DT_PROP_OR(DT_CHILD(node_id, channelb_0),          \
+							   sig_pos_mode, 0),                       \
+				.sig_neg_mode = DT_PROP_OR(DT_CHILD(node_id, channelb_0),          \
+							   sig_neg_mode, 0),                       \
+				.ctrl_l_mode =                                                     \
+					DT_PROP_OR(DT_CHILD(node_id, channelb_0), ctrl_l_mode, 0), \
+				.ctrl_h_mode =                                                     \
+					DT_PROP_OR(DT_CHILD(node_id, channelb_0), ctrl_h_mode, 0), \
+			},                                                                         \
+	},
+
+static struct pcnt_esp32_unit_config unit_config[] = {DT_INST_FOREACH_CHILD(0, UNIT_CONFIG)};
+
+static struct pcnt_esp32_config pcnt_esp32_config = {
+	.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0),
+	.clock_dev = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(0)),
+	.clock_subsys = (clock_control_subsys_t)DT_INST_CLOCKS_CELL(0, offset),
+	.unit_config = unit_config,
+	.unit_len = ARRAY_SIZE(unit_config),
+};
+
+static struct pcnt_esp32_data pcnt_esp32_data = {
+	.hal = {
+		.dev = (pcnt_dev_t *)DT_INST_REG_ADDR(0),
+	},
+	.cmd_sem = Z_SEM_INITIALIZER(pcnt_esp32_data.cmd_sem, 1, 1),
```

Comment:
Is this a binary semaphore? If so its intial value should be 0 no 1 and the number of keys should be 1 .

Reply:
Changed it to a mutex, thanks!